### ISSUE TITLE
is there a command line interface?

### ISSUE THREAD

**Issue by miltosc** [hsoft/dupeguru#288]
Can i use dupeguru from command line without the interface?
I wish it could run with switches from the command line with all the features of the GUI.
Is this possible or is it in your future plans?
Thank you for the great software.

**Comment by hsoft** [comment]
It's something I've been thinking about lately, but it would be to have more than command line. The results would need to be presented in a ncurses fashion as I don't see a workflow where simple text would be piped to another executable.

I mean, how should the dupe results be formatted, textually speaking, to show refs, dupes and all that so that it is usable by something else? I can't wrap my head around how it could possibly work, I'll need help thinking this through.

**Comment by Milkman00** [comment]
the function that I would like to be able to do is just compare two folders something simple for what I'm looking for would be great since I know the files its finding are exactly the same except for their filenames.

so a command as simple like:
dupeguru -ref="m:/downloads sorted" -norm="m:/downloads unsorted" send=Recyclebin

you could have the results printed to a txt file. And give the user options to show what he wants along side set of matches.

**Comment by Milkman00** [comment]
oh just had another idea to be able to create reference database file for a directory instead of having to scan all the files every time that directory is compared, and just compare the,few new files against the database.

**Comment by miltosc** [comment]
Personally i use it not only for deleting same files with different filenames.

One of its best features is to find different files (in filesize) with similar filenames.

My usually usage scenario is the following:

1)
I am setting up the preferences like the following to find videos and images usually with different size and similar filenames

scan type: filename 
filter hardness: 60
scan type: FILENAME
filter hardness: 60%
word weighting
match simmilar words
can mix title kind
ignore file smaller than: 100Kb

2)
Then i save the results

3)
I wrote a script that parses results and creates one html file that takes 2 screenshots of each video using ffmpeg and shows also other image files found by dupeguru.

4)
The html file has next of each thumbnail the filesize and a checkbox.
Then i check all the files that are really same since i can see the thumbnails.
Depending on file size sometimes i choose the big ones or the smaller files.
Finally i submit that html form and it creates a script with all the delete commands for the checked files.

(If you want i can send you that script)

Conclusion

So in my case i would like tou run dupeguru from command line to create the .dupeguru file using the preferences of the 1st step, 
then feed those results to my script and produce as a result the html file with the video and images previews to select which ones i want to delete.

I hope i was clear.
Since i do not know python (only php and shell script) i cannot say if it is easy or not to make dupeguru run from command line with these preferences as arguments.

Thanks again for the great software.
If i can help in any way tell me.

**Comment by hsoft** [comment]
If we leave out the responsibility of managing the results to the GUI and simply output search results in different formats (plain text dupe filename list, CSV, HTML, .dupeguru file), it makes the command line feature much more manageable to implement. I think we're getting somewhere.

As for enabling the direct deletion of dupes without reviews, it makes me nervous. It's usually a bad idea to delete dupes without review. If the output is plain text, it should be easy to simply pipe the results to `rm` if we need to.

**Comment by miltosc** [comment]
In my example my script it does not directly delete the files.
The user checks what he wants from my script's resulted html file by watching video and image thumbnails.

**Comment by hsoft** [comment]
@miltosc: I was referring to @Milkman00's comment who wrote that it's a feature he'd want.

**Comment by miltosc** [comment]
Ok. Sorry i misunderstood.

**Comment by Milkman00** [comment]
Yes,it can be dangerous to delete this way, and wouldn't be the first dangerous command to ever be implemented on a system,
In my scenario the files will have just been copied,renamed and sorted into folders using a batch file script I put together yesterday to use filebot (https://www.openhub.net/p/filebot), and simply moving the files instead can result in the unwanted deletion of files. 

I would be using a content scan in my scenario,and sending the files to the recyclebin(or a directory,one last review before they are deleted) 
A time stamped log of what has been deleted would be useful too.

**Comment by patrickatamaniuk** [comment]
cli first step could be scanning and comparing only, writing a loadable results file.
Second step could be firing up the details window from cli.
@Milkman00 the reference db, remembering what has been compared and to compare only new files gets a +1 here. Scanning is remembered in a database, but comparing currently is not, yet.

**Comment by hsoft** [comment]
@patrickatamaniuk I see from your different comments lately that you're interested in implementing incremental scanning :) I'm also +1 on it, so a quality PR of that sort will be merged. Don't hesitate to poke me if you need help navigating the code.

**Comment by patrickatamaniuk** [comment]
@hsoft yes. I am sifting through your (rather high quality) code right now, practicing objective c and playing around. Incremental scanning was done by T. Schmitt already at 1999, his picture comparison code (stic) is quite awesome, works much like yours. Commandline only, unix, BSD license. I need to wrap my head around both of your and his code, and there might be interesting results.

**Comment by patrickatamaniuk** [comment]
CLI Concept thoughts:
we got 2 major UI platforms: cocoa (mac) and qt (win/unix/linux)
Code-wise, cli feels like a third ui platform (ncurses or whatever)

But, i'd like have cli available for unix/linux, windows and mac (BSD) when i download the OS-specific dupeGuru package. So the build system would build 2 UI targets for each cocoa and qt: dupeGuru + dupeGuruCli.

The source would have a new tree 'cli' like cocoa and qt. It might contain all editions and all neccessery override-classes starting from app, all required view-like things and the job system.
I am thinking of 2 different views on a CLI version:
1. non-interactive. I do like to script things. (yes, yes, comparing dupes is interactive, and will be in the future. Scanning and comparing is not interactive, though)
2. interactive. Comparing files/images/music must stay interactive. Spawning comparison viewers or whatever could be a solution.

I believe the cli functionality should focus on non-interactive tasks, so that my home server farm can scan and match dupes, and the GUI version can read the results for human comparison and verification.

**Comment by jelabarre59** [comment]
Something slightly different from what you're thinking, but still command-line related.  I'd like to see command-line switches to let me give it the folder or folders to scan when the program loads.  I'm finding I'd like to parse through a particular folder or pair of folders first, and clear out the dups in smaller sets.  But doing this from the GUI means I have to drill down through the folders for each one to include, which is awkward and time-consuming, especially if the next folder(s) are parallel to the prior set (since I can't simply edit the prior set).  A bonus would be if I could say at the beginning which was the reference directory (again, I could use that on sub-sets of folders as I work my way through)
The problem I find is I either have to make a single large selection (and end up with a massive set of duplicates) or I have to drill down for each subset, even if they're nearly the same as what I already did, so that I can make a more manageable and smaller set of dupes to pick through.

**Comment by CyberPK** [comment]
why don't have dupeguru run an analysis outputting a .dupeguru file to be analyzed in a platform supporting the gui? Then the remote files could be "reconnected" and deleted "online" or can be created a ".delete" file to be taken in input in the original platform performing the selected action.
